storage.go

```diff
@@ -2,35 +2,268 @@ package azfile
+	}
+
+	count := int64(azfile.CountToEnd)
+	if opt.HasSize {
+		count = opt.Size
+	}
+
+	output, err := s.client.NewFileURL(path).Download(ctx, offset, count, false)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		cErr := output.Response().Body.Close()
+		if cErr != nil {
+			err = cErr
+		}
+	}()
+
+	rc := output.Response().Body
+	if opt.HasIoCallback {
+		rc = iowrap.CallbackReadCloser(rc, opt.IoCallback)
+	}
+
+	return io.Copy(w, rc)
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	var dirOutput *azfile.DirectoryGetPropertiesResponse
+	var fileOutput *azfile.FileGetPropertiesResponse
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		dirOutput, err = s.client.NewDirectoryURL(path).GetProperties(ctx)
+	} else {
+		fileOutput, err = s.client.NewFileURL(path).GetProperties(ctx)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	o = s.newObject(true)
+	o.ID = rp
+	o.Path = path
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		o.Mode |= ModeDir
+
+		o.SetLastModified(dirOutput.LastModified())
+
+		if v := string(dirOutput.ETag()); v != "" {
+			o.SetEtag(v)
+		}
+
+		var sm ObjectSystemMetadata
+		if v, err := strconv.ParseBool(dirOutput.IsServerEncrypted()); err == nil {
+			sm.ServerEncrypted = v
+		}
+		o.SetSystemMetadata(sm)
+	} else {
+		o.Mode |= ModeRead
+
+		o.SetContentLength(fileOutput.ContentLength())
+		o.SetLastModified(fileOutput.LastModified())
+
+		if v := string(fileOutput.ETag()); v != "" {
+			o.SetEtag(v)
+		}
+		if v := fileOutput.ContentType(); v != "" {
+			o.SetContentType(v)
+		}
+		if v := fileOutput.ContentMD5(); len(v) > 0 {
+			o.SetContentMd5(base64.StdEncoding.EncodeToString(v))
+		}
+
+		var sm ObjectSystemMetadata
+		if v, err := strconv.ParseBool(fileOutput.IsServerEncrypted()); err == nil {
+			sm.ServerEncrypted = v
+		}
+		o.SetSystemMetadata(sm)
+	}
+
+	return o, nil
 }
 
 func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
-	panic("not implemented")
+	if opt.HasIoCallback {
+		r = iowrap.CallbackReader(r, opt.IoCallback)
+	}
+
+	headers := azfile.FileHTTPHeaders{}
+
+	if opt.HasContentType {
+		headers.ContentType = opt.ContentType
+	}
+
+	// `Create` only initializes the file.
+	// ref: https://docs.microsoft.com/en-us/rest/api/storageservices/create-file
+	_, err = s.client.NewFileURL(path).Create(ctx, size, headers, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	body := iowrap.SizedReadSeekCloser(r, size)
+
+	var transactionalMD5 []byte
+	if opt.HasContentMd5 {
+		transactionalMD5, err = base64.StdEncoding.DecodeString(opt.ContentMd5)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	// Since `Create' only initializes the file, we need to call `UploadRange' to write the contents to the file.
+	_, err = s.client.NewFileURL(path).UploadRange(ctx, 0, body, transactionalMD5)
```

Comment:
What happened if `transactionalMD5` is empty?

Reply:
This is an optional value, so the null value does not affect the API call.